File: Algorithm/C++/96.unique-binary-search-trees.cpp
/*
 * @lc app=leetcode id=96 lang=cpp
 *
 * [96] Unique Binary Search Trees
 */

// @lc code=start
#include <vector>

class Solution {
public:
    int numTrees(int n) {
        std::vector<int> dp(n+1, 0);
        dp[0] = 1, dp[1] = 1;
        for(int i=2; i<n+1; i++) {
            for(int j=0; j<i; j++) {
                dp[i] += dp[j]*dp[i-j-1];
            }
        }
        return dp[n];
    }
};
// @lc code=end

File: Algorithm/Python/96.unique-binary-search-trees.py
#
# @lc app=leetcode id=96 lang=python
#
# [96] Unique Binary Search Trees
#

# @lc code=start
class Solution(object):
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [0] * (n+1)
        dp[0], dp[1] = 1, 1
        for i in range(2, n+1):
            for j in range(i):
                dp[i] += dp[j]*dp[i-j-1]
        return dp[n]
        
# @lc code=end

